File: apps/desktop/src-tauri/src/commands/completions.rs
use tauri::State;

use crate::errors::Result;
use crate::keystore::KeystoreState;
use crate::shell_completions::ShellCompletionsManager;
use crate::storage::command_history::{self, CommandHistoryEntry};
use crate::AppState;

/// Record one executed command line for a scope. Filtered lines (leading space,
/// secret-bearing, empty, oversized) return `false` and store nothing.
#[tauri::command]
pub async fn command_history_record(
    state: State<'_, AppState>,
    scope_key: String,
    command: String,
) -> Result<bool> {
    command_history::record(&state.pool, &scope_key, &command).await
}

/// Prefix-matched history for a scope, most-used first.
#[tauri::command]
pub async fn command_history_query(
    state: State<'_, AppState>,
    scope_key: String,
    prefix: String,
    limit: Option<i64>,
) -> Result<Vec<CommandHistoryEntry>> {
    command_history::query(&state.pool, &scope_key, &prefix, limit.unwrap_or(20)).await
}

/// Names on the remote host's `$PATH`, for first-token completion (SSH only).
#[tauri::command]
pub async fn shell_completions_executables(
    state: State<'_, AppState>,
    manager: State<'_, ShellCompletionsManager>,
    keystore_state: State<'_, KeystoreState>,
    host_id: String,
) -> Result<Vec<String>> {
    manager
        .executables(&state.pool, &keystore_state, &host_id)
        .await
}

/// Entries of one remote directory, for path-token completion (SSH only).
/// Directory entries carry a trailing `/`.
#[tauri::command]
pub async fn shell_completions_paths(
    state: State<'_, AppState>,
    manager: State<'_, ShellCompletionsManager>,
    keystore_state: State<'_, KeystoreState>,
    host_id: String,
    dir: String,
) -> Result<Vec<String>> {
    manager
        .paths(&state.pool, &keystore_state, &host_id, &dir)
        .await
}

File: apps/desktop/src/lib/completions.ts
import { invoke } from "@tauri-apps/api/core";

/** One remembered command line for a scope. */
export type CommandHistoryEntry = {
  command: string;
  useCount: number;
  lastUsedAt: number;
};

/**
 * Record an executed command line. Returns whether it was stored — the backend
 * silently drops lines that start with a space, look secret-bearing, or are
 * empty/oversized.
 */
export function recordCommandHistory(scopeKey: string, command: string): Promise<boolean> {
  return invoke<boolean>("command_history_record", { scopeKey, command });
}

/** Prefix-matched history for a scope, most-used first. */
export function queryCommandHistory(
  scopeKey: string,
  prefix: string,
  limit?: number,
): Promise<CommandHistoryEntry[]> {
  return invoke<CommandHistoryEntry[]>("command_history_query", {
    scopeKey,
    prefix,
    limit: limit ?? null,
  });
}

/** Names on the remote host's `$PATH` (SSH hosts only). */
export function fetchRemoteExecutables(hostId: string): Promise<string[]> {
  return invoke<string[]>("shell_completions_executables", { hostId });
}

/** Entries of one remote directory; directories carry a trailing `/`. */
export function fetchRemotePaths(hostId: string, dir: string): Promise<string[]> {
  return invoke<string[]>("shell_completions_paths", { hostId, dir });
}
